### ISSUE TITLE
stamp: scaling

### ISSUE THREAD

**Issue by maxiride** [pdfcpu/pdfcpu#425]
I thought to have understood the stamp command and the scaling option but I'm encountering this issue:
I have a 5x25mm — 300dpi — 60x296px — jpg file, I want to stamp it on a A4 pdf and the stamp on the document must have the exact dimensions of 5x25mm.

I thought I had to use the option `sc: 1 abs` but the resulting PDF has an huge stamp. I am now tweaking the scaling by trial and error to workaround the issue, where did I misunderstood the documentation?

File being stamped.
![11_300](https://user-images.githubusercontent.com/3630857/152558931-8fa2021b-81e0-4a58-bd22-ec59a66a9fac.png)

**Comment by hhrutter** [comment]
Hi there!

Is this happening on the CLI or API?
Can you share details about this specific stamping operation and share the parameters?
Any rotation involved?
You may have encountered a bug there.

In general if you stamp an image all you have is an image with its dimensions in pixels: width and height.
Anything beyond that is up to the renderer.

**Comment by maxiride** [comment]
> Can you share details about this specific stamping operation and its parameters?

The file applied is the one attached in the first comment and the parameters of the operation are the following:

    pdfcpu stamp add -m image 11.jpg "sc:1 abs, rot:0" in.pdf out.pdf


The image, as mentioned, is 60x296px and it is saved at 300dpi  which should render as a 5x25mm.

> In general if you stamp an image all you have is an image with its dimensions in pixels: width and height.
Anything beyond that is up to the renderer.

As with any media for printing shouldn't the DPI matter too? The same resolution (in pixel) yields very different final sizes at different DPIs. Translating pixels to millimetres can't be done without factoring in the dpi.

Maybe I'm getting lost in the details here and the issue is on the reverse, at which resolution (px width and height) and dpi should one create an image to be stamped so that it is applied with a target size (in mm)?

**Comment by maxiride** [comment]
I suppose the area of intervention might reside here?

https://github.com/pdfcpu/pdfcpu/blob/8c35cdc14a868414ceb195dca0ef9535ef472b76/pkg/pdfcpu/readImage.go#L415-L419

**Comment by hhrutter** [comment]
You need to treat the image(width,height) separately from the destination DPI.
(which really equals to the destinations pixels per inch).

The DPI in effect are controlled by the PDF user units defined for the PDF user space, which defaults to 72.

The import command already supports a desired DPI to apply for an image, but here this amounts to a straightforward scaling because images get rendered on empty pages using the default user units (=72dpi).

During stamping - different story. One needs to take into account the user units in effect before rendering an image
and then apply proper scaling - **and any transformation in effect**.

In a nutshell you are right - Desired DPI is a missing feature for stamping but what you stumbled upon may very well be just a bug. I am going to take a look at your sample and take it from there.

**Comment by hhrutter** [comment]
You are working on wrong assumptions:

If you render an image 1:1 (absolute scaling 1.0) then the aspect ratio (=width/height) of your image will be preserved.
That means you will never get to 5x25mm neither with 50x296 nor a 118x591 image.

What we really need here is something the create via json command [already supports](https://github.com/pdfcpu/pdfcpu/blob/master/pkg/samples/create/images.pdf):
Specify your destination *width* or *height* or both and apply a best fit approach.
..or specify some virtual destination DPI for scaling but I am now more leaning to width/height because that gives you more control.

I will look into adding these to the stamp command. In addition to what we really need - and the create command is not supporting it as of now - is also specifying the desired width/height user units.

So we need the desired user units both for create and stamp.
Lots of work ahead..

**Comment by maxiride** [comment]
Given all that you said I suppose that this ticket can be closed (I leave that to you if you deem it right). It's either a matter of using a more sophisticated command (through the use of the JSON input) and\or just provide images taking in account 72 dpi.
